src/main.rs

```diff
@@ -0,0 +1,116 @@
+// Copyright 2020 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::ffi::{OsStr, OsString};
+use std::io;
+
+use clap::{crate_version, App, AppSettings, Arg};
+use error_chain::quick_main;
+use users::os::unix::UserExt;
+use users::switch::{switch_user_group, SwitchUserGuard};
+use users::{get_user_by_name, User};
+
+use crate::errors::{Result, ResultExt};
+use crate::keys::read_keys;
+
+mod errors;
+mod keys;
+
+quick_main!(run);
+
+struct Config {
+    username: OsString,
+}
+
+fn parse_args() -> Result<Config> {
+    // Args are listed in --help in the order declared here.  Please keep
+    // the entire help text to 80 columns.
+    let matches = App::new("ssh-key-dir")
+        .version(crate_version!())
+        .global_setting(AppSettings::UnifiedHelpMessage)
+        .about("Print SSH keys from a user's ~/.ssh/authorized_keys.d")
+        .arg(
+            Arg::with_name("user")
+                .help("Username of connecting user")
+                .required(true)
+                .takes_value(true),
+        )
+        .get_matches();
+
+    Ok(Config {
+        username: matches
+            .value_of_os("user")
+            .map(OsString::from)
+            .expect("username missing"),
+    })
+}
+
+fn switch_user(username: &OsStr) -> Result<(User, SwitchUserGuard)> {
+    let user = get_user_by_name(username).chain_err(|| format!("no such user {:?}", username))?;
+    let guard = switch_user_group(user.uid(), user.primary_group_id())
+        .chain_err(|| "couldn't switch user")?;
+    Ok((user, guard))
+}
+
+fn run() -> Result<()> {
+    // parse args
+    let opts = parse_args()?;
+
+    // switch user
+    let (user, _switch_guard) = switch_user(&opts.username)?;
```

Comment:
Reading the docs on `switch_user_group`, it says that one should explicitly call `drop()` on the guard. Let's do that here before returning `Ok(())` so that if this code is ever refactored, we make sure that it comes along?